pkg/roachprod/cloud/cluster_cloud.go

```diff
@@ -346,14 +355,44 @@ func CreateCluster(l *logger.Logger, opts []*ClusterCreateOpts) error {
 			p = (p + 1) % providerCount
 		}
 
+		var vms vm.List
+		var vmsLock syncutil.Mutex
+		// Create VMs in parallel across all providers.
+		// Each provider will return the list of VMs it created, and we append
+		// them to the cached Cluster.
 		if err := vm.ProvidersParallel(o.CreateOpts.VMProviders, func(p vm.Provider) error {
-			return p.Create(l, vmLocations[p.Name()], o.CreateOpts, o.ProviderOptsContainer[p.Name()])
+			vmList, err := p.Create(l, vmLocations[p.Name()], o.CreateOpts, o.ProviderOptsContainer[p.Name()])
+			if err != nil {
+				return err
+			}
+			vmsLock.Lock()
+			defer vmsLock.Unlock()
+			vms = append(vms, vmList...)
+			return nil
 		}); err != nil {
-			return err
+			return c, err
 		}
+
+		c.VMs = append(c.VMs, vms...)
 	}
 
-	return nil
+	// As VMs are created in parallel, sort them to ensure consistent ordering.
+	// SyncedCluster.Wait() tends to be flaky if the VMs are not sorted and VMs
+	// become available in an unexpected order.
+	// Sort is also used in ListCloud() to ensure consistent ordering.
+	sort.Sort(c.VMs)
```

Comment:
The fact that `c.VMs` are expected to be in sorted (lexicographic) order is actually a hidden, correctness assumption. It's unfortunate that it's not documented anywhere else in `roachprod` code. We can fix that by paraphrasing the above, as well as, `ListCloud` where `Sort` is also used. Something to the effect of, "`roachprod.Start` expects nodes/vms to be in sorted order` would suffice. For further details, we can refer the reader to [1].

[1] https://github.com/cockroachdb/cockroach/pull/133647

Reply:
Added comment to each `Sort` and reference to PR in `ListCloud`.